src/ds5/ds5-thermal-handler.cpp

```diff
@@ -0,0 +1,188 @@
+        {
+            if (auto s = dynamic_cast<ds5_recalibrable_color_sensor*>(&(dev.get_sensor(i))))
+            {
+                _recalib_sensor = std::dynamic_pointer_cast<ds5_recalibrable_color_sensor>(s->shared_from_this());
+                break;
+            }
+        }
+
+        _tl_activation = std::make_shared<uvc_xu_option<uint8_t>>(dynamic_cast<uvc_sensor&>(*activation_sensor.get_raw_sensor()),
+            ds::depth_xu, ds::DS5_THERMAL_COMPENSATION, "Toggle Thermal Compensation Mechanism");
+
+        try {
+            _control_on = static_cast<bool>(_tl_activation->query());
+        }
+        catch (...)
+        {
+            _control_on = true;
+            LOG_WARNING("Initial TL Control state could not be verified, assume on ");
+        }
+
+        activation_sensor.register_option(RS2_OPTION_THERMAL_COMPENSATION, std::make_shared<thermal_compensation>(this));
+    }
+
+    ds5_thermal_handler::~ds5_thermal_handler()
+    {
+        if (_streaming_on && _control_on)
+            stop();
+    }
+
+    void ds5_thermal_handler::start()
+    {
+        _active_object.start();
+    }
+
+    void ds5_thermal_handler::stop()
+    {
+        _active_object.stop();
+        _temp_base = 0.f;
+
+        // Enforce calibration re-read on deactivation
+        if (auto sp = _recalib_sensor.lock())
+            sp->reset_calibration();
+
+        notify_of_calibration_change(RS2_CALIBRATION_SUCCESSFUL);
+    }
+
+
+    void ds5_thermal_handler::trigger_device_calibration(rs2_calibration_type type)
+    {
+        bool change_required = false;
+        bool activate = (RS2_CALIBRATION_THERMAL == type);
+        if (_streaming_on != activate)
+        {
+            if (auto sp = _dpt_sensor.lock())
+            {
+                // Thermal Activation occurs on depth sensor streaming changes
+                change_required = (sp->is_opened()) && (!sp->is_streaming());
+            }
+            if (change_required)
+            {
+                LOG_DEBUG("Thermal compensation polling was changed from " << (int)_streaming_on << " to " << (int)activate);
+                _streaming_on = activate;
+                update_mode(true);
+            }
+        }
+    }
+
+    void ds5_thermal_handler::set_feature(bool state)
+    {
+        bool change_required = false;
+        LOG_DEBUG(__FUNCTION__ << " state = " << state << " _control_on = " << _control_on);
+        if (state != _control_on)
+        {
+            _tl_activation->set(state);
+            _control_on = state;
+            update_mode();
+        }
+    }
+
+    float ds5_thermal_handler::query()
+    {
+        auto ctrl_state = static_cast<bool>(_tl_activation->query());
+        LOG_DEBUG(__FUNCTION__ << " _control_on = " << _control_on << " ctrl_state = " << ctrl_state);
+        if (_control_on != ctrl_state)
+        {
+            _control_on = ctrl_state;
+            update_mode();
+        }
```

Comment:
These operations are not expected when seeing the method's name - please consider to modify it.